priority-channel/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use std::{
+	pin::Pin,
+	result,
+	task::{Context, Poll},
+};
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+use async_channel::bounded;
+pub use async_channel::{TryRecvError, TrySendError};
+use futures::Stream;
+
+/// Create a wrapped `mpsc::broadcast` pair of `Sender` and `Receiver`.
+pub fn channel<T>(capacity: usize) -> (Sender<T>, Receiver<T>) {
+	let (tx, rx) = bounded::<T>(capacity);
+	let (tx_priority, rx_priority) = bounded::<T>(capacity);
+	(Sender { inner_priority: tx_priority, inner: tx }, Receiver { inner_priority: rx_priority, inner: rx })
+}
+
+/// A receiver tracking the messages consumed by itself.
+#[derive(Debug)]
+pub struct Receiver<T> {
+	inner: async_channel::Receiver<T>,
+	inner_priority: async_channel::Receiver<T>,
+}
+
+impl<T> std::ops::Deref for Receiver<T> {
+	type Target = async_channel::Receiver<T>;
+	fn deref(&self) -> &Self::Target {
+		&self.inner
+	}
+}
+
+impl<T> std::ops::DerefMut for Receiver<T> {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		&mut self.inner
+	}
+}
+
+impl<T> Stream for Receiver<T> {
+	type Item = T;
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		match async_channel::Receiver::poll_next(Pin::new(&mut self.inner_priority), cx) {
+			Poll::Ready(maybe_value) => Poll::Ready(maybe_value),
+			Poll::Pending => match async_channel::Receiver::poll_next(Pin::new(&mut self.inner), cx) {
+				Poll::Ready(maybe_value) => Poll::Ready(maybe_value),
+				Poll::Pending => Poll::Pending,
+			},
+		}
+	}
+
+	fn size_hint(&self) -> (usize, Option<usize>) {
+		self.inner.size_hint()
+	}
+}
+
+impl<T> Clone for Receiver<T> {
+	fn clone(&self) -> Self {
+		Self { inner: self.inner.clone(), inner_priority: self.inner_priority.clone() }
+	}
+}
+
+impl<T> Receiver<T> {
+	/// Attempt to receive the next item.
+	pub fn try_next(&mut self) -> Result<T, TryRecvError> {
+		match self.inner_priority.try_recv() {
```

Comment:
I don't think it is a good idea. Priority must not be misused but adding sampling won't really solve any real issue. If we overflow priority queue, the non-priority queue will always be penalized. TBH, I was thinking exactly about the same idea when doing that design, but I have rejected it due to complexity and difficulty in understanding. The rule of thumb is that high priority channels must not have bulk messages.

Reply:
That makes sense and I agree on when we should use priority channels or not. However what it is not take into account is what happens under high load, when CPU resources are less than what is needed. I don't think it is necessarily complex or hard to understand since this kind of problem is common.

Reply:
It is now possible to select a strategy if needed.

Reply:
Thanks, this looks neat!